File: 6주차/LeetCode_46_Permutations/dongjun.java
/**
 *  @문제접근
 *  완전탐색으로 풀었습니다.
 *
 *  @성능
 *  Runtime: 1 ms
 *  Memory Usage: 40 MB
 */
public class Permutations {
    class Solution {
        public static boolean[] visited;
        public static List<List<Integer>> ret;

        public List<List<Integer>> permute(int[] nums) {
            ret = new ArrayList<>();
            visited = new boolean[nums.length];
            LinkedList<Integer> start = new LinkedList<>();

            dfs(start, nums);

            return ret;
        }

        public void dfs(LinkedList<Integer> now, int[] nums){
            if(now.size() == nums.length){
                ret.add(new LinkedList<>(now));
                return;
            }

            for(int i = 0; i < nums.length; i++){
                if(!visited[i]){
                    visited[i] = true;
                    now.add(nums[i]);
                    dfs(now, nums);
                    now.pollLast();
                    visited[i] = false;
                }
            }
        }
    }
}

File: 8주차/LeetCode_283_Move_Zeroes/dongjun.py
"""
1. 문제접근
	리스트를 순회하며, 0을 만나면 삭제하고
	리스트 뒤에 0을 삽입했습니다.

2. 성능
	Runtime: 44ms
	Memory Usage: 15.1MB
	시간복잡도: O(N)
"""
from collections import deque

class Solution:
    def moveZeroes(self, nums: List[int]) -> None:

        ret = nums.copy()
        length = len(nums)
        cnt = 0
        for i in range(length):
            if ret[i] == 0:
                del nums[i-cnt]
                cnt += 1
                nums.append(0)

        return nums
